### PR TITLE
Raw-quoting syntax

### PR DESCRIPTION
**Pull request by krobelus**
TL;DR: this makes it so that typing `(` at the start of a command opens a subshell;
completions, quoting etc. work just like inside command substitutions.
This is enabled by the first commit that adds raw quoting syntax.

---

Other shells allow to create subshells using `(cd somewhere && do-something)`.
Our answer to this is to create a subshell like a regular process (#1439):

	$SHELL -c 'cd somewhere && do-something'

This syntax works the same way for non-shell programs

	watch 'ps aux | grep foo'
	git bisect run $SHELL -c 'make || exit 125; make test'
	hyperfine '...'

Issues with our solution are:
- the argument has no command-highlighting, completions (#9406) or indentation
- the  user may need to escape the argument multiple times ("quoting hell").
This primarily affects arguments containing backslashes and quotes
(for example `$SHELL -c 'printf %s\\\n *'`).
In many other cases backslash escaping is not necessary however the rules
can be confusing.

One workaround is to compose the subcommand in a separate shell prompt and
copy-paste it (inside quotes for automatic escaping).
That's not ideal because it breaks the writer's flow, and it doesn't work
well if the command needs to be refined.

Add a `(( ))` raw quoting syntax for subshells and deferred commands.
Syntax highlighting, completion and indentation work the same as in top-level
commands. No expansion/unescaping takes place:

	$SHELL -c ((printf %s\\n *))

Nested parentheses are okay, similar to Tcl's braced strings.
Imbalanced quotation marks are an error.
This means that the quoted string cannot contain an imbalance of its quote
characters.
(The good side is that, in many cases, imbalance is not intentional).

If we stick to this syntax, it would be good to also support `(< >)` and
`({ })` with the same meaning as `(( ))`, to make it easier to use when the
quoted content contains an unmatched `(`.

---

Also add a `{{ }}` raw quoting syntax that does not get shell highlighting,
completion and indentation.
This is meant for languages like Perl, awk, Python and regex dialects (no
more need to escape backslashes).
The would be a breaking change, we might want to avoid that by picking a
different syntax.

To-do:
- figure out how we want to tweak syntax, possibly find something that can
be added to Bash.
- adjust __fish_tokenizer_state
- docs/changelog/tests

### CODE REVIEW

**Comment by faho** [comment]
Note: I changed the title to avoid confusing people. "Subshell" is a very specific thing, this doesn't really add it, and we've overused the term in the past.

----

My first instinct here is that `((` does not strike me obvious as quoting, and it is not familiar to people coming from bash or python.

We've brought up `"""` python-style quoting before, I feel like that's more intuitive.

There is also the question of how to deal with indentation in quotes - one problem with embedding python scripts specifically is this:

```fish
if foo
    python -c "print('foo')
print('bar')" # <- this *cannot* be indented or it breaks python
end
```


I also am not a fan of deferring to `$SHELL` in the keybinding, which can be any sort of shell, and often is not fish. To be honest, I would remove the binding.

**Comment by ridiculousfish** [comment]
Agreed that we should not promote the "run fish recursively" psuedo-subshell in key bindings. We'll have a real solution for that which preserves variables, functions, etc. which would be incompatible with invoking $SHELL.

+1 to `"""` for raw strings.

**Comment by krobelus** [comment]
Does this mean we agree or disagree on adding a command-flavored quoting syntax to enable
`with-cd ./office (( do work ))`?

**Comment by faho** [comment]
> we agree or disagree on adding a command-flavored quoting syntax

As I understand it, it has the exact same semantics, language-wise, it's only different interactively in that it will tell the reader "this in here is a command, highlight it like one"?

I'm not really sold on it, I don't think fish-script-inside-of-fish-script is a common enough thing to add more syntax for.

Specifically for some cases like `fish -c ... &` to background scripts we'd be better off just allowing blocks/functions to be backgrounded (which, granted, is a bigger task) than add syntax to make the workaround nicer *interactively*.

And again the symbols are alien (`(())` is arithmetic expansion in bash) except if you're used to TCL, which I don't have fond memories of (it's a big part of why we moved off of expect to pexpect), and which I don't think is very popular. `{{}}` is probably a bit worse because there's more backwards-incompatibility there, but I don't like `(())` as a choice either. Funnily enough `(<>)` would at least be a syntax error at the moment (also in bash, so there's less chance for confusion).

What moving to `"""` and `'''` would gain us is more familiar quoting (it's in python) and an obvious way to have "quoting with less escaping" but still interpolate variables, which seems like an important part.

**Comment by krobelus** [comment]
actually for my use case plain quoting is enough. Raw quoting is `r""`.

**Comment by krobelus** [comment]
The primary and concrete problem I'm trying to solve is to quickly run some commands in a different directory.
I don't know how to do this in a nice way while making sure that `$PWD` is restored on `ctrl-c`.
I believe this is an important scenario.

So far I've been using `sh -c 'cd ... && ...'`.
The immediate problem with that is that completion is not yet implemented.
I have an implementation that is almost finished, so that should cover many simple cases.
(As you noted, it has false positives for other shells but I think we can accept that as an intermediate step?)

I think there's a reasonably good chance that we want something better (native, no quoting hell etc.)

Raw quoting / deferred commands is one approach. It's definitely not the only one, and I'd love to hear other ideas such as

```
cd --scoped ./somewhere
    echo do something somewhere
    echo do something else
end
```

There's also some ideas leveraging `begin/end` (like `cd -l`) that would work well if they could be written inside `{}` or `()` but that's probably not an easy sell either.

I think that we should probably figure that out before we talk about a concrete quoting syntax.
For that there are countless possibilities, I'm sure we can find something that does not mislead users.
However, if the syntax could hold a fish script payload I expect that it will be foolish to not support trivial nesting.

Raw quoting for generic strings (regex etc) is also worthwhile exploring but it's less related to my problem and I haven't yet found real-world use case where it helps.

> """ still interpolate variables, which seems like an important part.

For the generic one maybe but it's probably less needed for quoted commands (`--no-scope-shadowing` is a rare exception).

**Comment by faho** [comment]
> So far I've been using sh -c 'cd ... && ...'.
The immediate problem with that is that completion is not yet implemented.

Of course now you're in sh-land, and fish won't really ever be able to complete sh-commands.

But it's also just... a much bigger tool than you would need for the idea at hand.

> The primary and concrete problem I'm trying to solve is to quickly run some commands in a different directory.

I don't know why you would need a special quoting syntax for that, it seems like the right solution for this is something much more focused, like allowing $PWD to be set locally, so you can do `PWD=/some/dir; begin foo; end`.

Even if you *did* want a full "subshell" experience, that could just be something like `begin --subshell` - with no extra quoting *at all*, and no awkward switching into an entire different shell and language.

>  However, if the syntax could hold a fish script payload I expect that it will be foolish to not support trivial nesting.

I don't believe quoted fish-script inside of quoted fish-script is something we really need to be concerned with much. I don't believe it's a common thing, and when it comes up you could always just refactor (e.g. make multiple scripts).

**Comment by krobelus** [comment]
> I don't know why you would need a special quoting syntax for that,

I did not say that.

> it seems like the right solution for this is something much more focused,
> like allowing $PWD to be set locally, so you can do PWD=/some/dir; begin
> foo; end.

`PWD=some-dir foo` sounds good to me. Discoverability will be worse but so be it

> and no awkward switching into an entire different shell and language.

?
sh is just an example, could be proper fish.
I don't think that playing nice with the outside world is a bad thing.

> I don't believe quoted fish-script inside of quoted fish-script is something
> we really need to be concerned with much. I don't believe it's a common
> thing, and when it comes up you could always just refactor (e.g. make
> multiple scripts).

you're repeating yourself / stating the obvious

**Comment by krobelus** [comment]
how about `{}` or `()` for command grouping then?

**Comment by faho** [comment]
Not possible considering backwards-compatibility.

I also don't like having `{}` pull double-duty like that.

**Comment by krobelus** [comment]
`()` does not have a backwards compatibility issue
